libcaf_net/caf/net/web_socket/framing.cpp

```diff
@@ -63,119 +78,167 @@ void framing::abort(const error& reason) {
       return -1;
     }
   } else {
     // Make sure the entire frame (including header) fits into max_frame_size.
-    if (hdr.payload_len >= max_frame_size - static_cast<size_t>(hdr_bytes)) {
+    if (hdr_.payload_len >= max_frame_size - static_cast<size_t>(hdr_bytes)) {
       CAF_LOG_DEBUG("WebSocket frame too large");
       abort_and_shutdown(sec::protocol_error, "WebSocket frame too large");
       return -1;
     }
     // Reject any payload that exceeds max_frame_size. This covers assembled
-    //  payloads as well by including payload_buf_.
-    if (payload_buf_.size() + hdr.payload_len > max_frame_size) {
+    // payloads as well by including payload_buf_.
+    if (payload_buf_.size() + hdr_.payload_len > max_frame_size) {
       CAF_LOG_DEBUG("fragmented WebSocket payload exceeds maximum size");
       abort_and_shutdown(sec::protocol_error, "fragmented WebSocket payload "
                                               "exceeds maximum size");
       return -1;
     }
   }
-  // Wait for more data if necessary.
-  size_t frame_size = hdr_bytes + hdr.payload_len;
-  if (buffer.size() < frame_size) {
-    down_->configure_read(receive_policy::exactly(frame_size));
-    return 0;
-  }
-  // Decode frame.
-  auto payload_len = static_cast<size_t>(hdr.payload_len);
-  auto payload = buffer.subspan(hdr_bytes, payload_len);
-  if (hdr.mask_key != 0) {
-    detail::rfc6455::mask_data(hdr.mask_key, payload);
+  return hdr_bytes;
+}
+
+ptrdiff_t framing::consume_payload(byte_span buffer, byte_span delta) {
+  // Calculate at what point of the received buffer the delta payload begins.
+  auto offset = static_cast<ptrdiff_t>(buffer.size() - delta.size());
+  // We read frame_size at most. This can leave unprocessed input.
+  auto payload = buffer.first(std::min(buffer.size(), hdr_.payload_len));
+  // Unmask the arrived data.
+  if (hdr_.mask_key != 0)
+    detail::rfc6455::mask_data(hdr_.mask_key, payload, offset);
+  // In case of text message, we want to validate the UTF-8 encoding early.
+  if (hdr_.opcode == detail::rfc6455::opcode_type::text_frame
+      || (hdr_.opcode == detail::rfc6455::opcode_type::continuation_frame
+          && opcode_ == detail::rfc6455::opcode_type::text_frame)) {
+    if (hdr_.opcode == detail::rfc6455::opcode_type::text_frame && hdr_.fin) {
+      if (!payload_valid(payload, validation_offset_)) {
+        abort_and_shutdown(sec::malformed_message, "invalid UTF-8 sequence");
+        return -1;
+      }
+    } else {
+      payload_buf_.insert(payload_buf_.end(), payload.begin() + offset,
+                          payload.end());
+      if (!payload_valid(payload_buf_, validation_offset_)) {
+        abort_and_shutdown(sec::malformed_message, "invalid UTF-8 sequence");
+        return -1;
+      }
+    }
+    // Wait for more data if necessary.
+    if (buffer.size() < hdr_.payload_len) {
+      down_->configure_read(receive_policy::up_to(hdr_.payload_len));
```

Comment:
Again, I would expect that the `consume_header` function does this and we simply return `0` here until all data arrived.

Reply:
:+1: